internal/statesync/reactor.go

```diff
@@ -371,7 +367,9 @@ func (r *Reactor) backfill(
 				case height := <-queue.nextHeight():
 					peer := r.peers.Pop(ctx)
 					r.Logger.Debug("fetching next block", "height", height, "peer", peer)
-					lb, err := r.dispatcher.LightBlock(ctxWithCancel, height, peer)
+					subCtx, cancel := context.WithTimeout(ctxWithCancel, lightBlockResponseTimeout)
+					defer cancel()
+					lb, err := r.dispatcher.LightBlock(subCtx, height, peer)
```

Comment:
love this change!

Reply:
I agree. My only remaining concern is that a `defer` inside a loop stacks up state until the enclosing function returns. We can hack that with a closure (destructors, Go style), e.g.,

```go
subctx, cancel := context.WithTimeout(ctx, lightBlockResponseTimeout)
lb, err := func() (*types.LightBlock, error) {
  defer cancel()
  return r.dispatcher.LightBlock(subctx, height, peer)
}()
```

or by explicitly calling it after the call but before we check the errors. The latter is actually easier, but the attraction of a defer for happening even in case of a panic is understandable.